File: src/ViewModels/Sakuno.ING.ViewModels.Logging/FilterVM.vb
Public Class FilterVM(Of T)
    Inherits BindableObject
    Implements IFilterVM
    Implements IEqualityComparer(Of T)

    Private ReadOnly keySelector As Func(Of T, Integer)
    Private ReadOnly defaultText As Func(Of T, String)
    Private ReadOnly acceptingText As Func(Of T, String())

    Public Sub New(name As String, keySelector As Func(Of T, Integer), defaultText As Func(Of T, String), Optional acceptingText As Func(Of T, String()) = Nothing)
        Me.Name = name
        Me.keySelector = keySelector
        Me.defaultText = defaultText
        Me.acceptingText = acceptingText
    End Sub

    Public Function Hits(value As T) As Boolean
        If Not IsEnabled Then Return True
        If defaultText(value).Contains(SelectedText) Then Return True
        If acceptingText Is Nothing Then Return False
        For Each str In acceptingText(value)
            If str.Contains(SelectedText) Then Return True
        Next
        Return False
    End Function

    Public ReadOnly Property Name As String Implements IFilterVM.Name

    Private _isEnabled As Boolean
    Public Property IsEnabled As Boolean Implements IFilterVM.IsEnabled
        Get
            Return _isEnabled
        End Get
        Set(value As Boolean)
            [Set](_isEnabled, value)
        End Set
    End Property
    Public Property SelectedText As String Implements IFilterVM.SelectedText

    Private _candidates As IBindableCollection(Of String)
    Public ReadOnly Property Candidates As IBindableCollection(Of String) Implements IFilterVM.Candidates
        Get
            Return _candidates
        End Get
    End Property
    Public Sub UpdateCandidates(values As IEnumerable(Of T))
        _candidates = values.Distinct(Me).OrderBy(keySelector).Select(defaultText).Distinct().ToBindable()
        NotifyPropertyChanged(NameOf(Candidates))
    End Sub

    Private Overloads Function Equals(x As T, y As T) As Boolean Implements IEqualityComparer(Of T).Equals
        Return keySelector(x) = keySelector(y)
    End Function

    Private Overloads Function GetHashCode(obj As T) As Integer Implements IEqualityComparer(Of T).GetHashCode
        Return keySelector(obj)
    End Function
End Class

File: src/ViewModels/Sakuno.ING.ViewModels.Logging/FilterVM.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Sakuno.ING.ViewModels.Logging
{
    internal class FilterVM<T> : BindableObject, IFilterVM, IEqualityComparer<T>
    {
        private readonly Func<T, int> keySelector;
        private readonly Func<T, string> defaultText;
        private readonly Func<T, string[]> acceptingText;

        public FilterVM(string name, Func<T, int> keySelector, Func<T, string> defaultText, Func<T, string[]> acceptingText = null)
        {
            this.keySelector = keySelector;
            this.defaultText = defaultText;
            this.acceptingText = acceptingText;
            Name = name;
        }

        public string Name { get; }

        public bool Hits(T value)
        {
            if (defaultText(value).Contains(SelectedText)) return true;
            if (acceptingText == null) return false;
            foreach (string str in acceptingText(value))
                if (str.Contains(SelectedText)) return true;

            return false;
        }

        private string _selectedTest;
        public string SelectedText
        {
            get => _selectedTest;
            set
            {
                if (_selectedTest != value)
                {
                    _selectedTest = value;
                    NotifyPropertyChanged();
                    Updated?.Invoke();
                }
            }
        }

        private bool _isEnabled;
        public bool IsEnabled
        {
            get => _isEnabled;
            set
            {
                if (_isEnabled != value)
                {
                    _isEnabled = value;
                    NotifyPropertyChanged();
                    Updated?.Invoke();
                }
            }
        }

        public event Action Updated;

        public IBindableCollection<string> Candidates { get; private set; }
        public void UpdateCandidates(IEnumerable<T> values)
        {
            Candidates = values.Distinct(this).OrderBy(keySelector)
                .Select(defaultText).Distinct().ToBindable();
            NotifyPropertyChanged(nameof(Candidates));
        }

        bool IEqualityComparer<T>.Equals(T x, T y) => keySelector(x) == keySelector(y);
        int IEqualityComparer<T>.GetHashCode(T obj) => keySelector(obj);
    }
}
